include/tvm/relax/nested_msg.h

```diff
@@ -341,6 +370,45 @@ void DecomposeNestedMsg(Expr expr, NestedMsg<T> msg, FType fvisitleaf) {
   }
 }
 
+/*!
+ * \brief Recursively transform the tuple structure in expr and msg along with it.
+ *
+ * This function will call fvisitleaf for each leaf expression in expr.
+ * This function will throw an error if the nesting structure in msg does not
+ * match the tuple nesting structure in expr.
+ *
+ * \param expr The input expression to be transform.
+ * \param msg The input message to guide the transformation.
+ * \param fvisitleaf with signature fvisitleaf(Expr expr, NestedMsg<T> msg)->Expr
+ * \tparam T the content type of nested msg
+ * \tparam FType The visit function type.
+ */
+template <typename T, typename FType>
+Expr TransformTupleLeaf(Expr expr, NestedMsg<T> msg, FType fvisitleaf) {
+  StructInfo sinfo = GetStructInfo(expr);
+  if (const auto* tuple = sinfo.as<TupleStructInfoNode>()) {
+    ICHECK(msg.IsNested()) << "Expected nested to match tuple";
+    bool same = true;
+    Array<Expr> fields;
+    fields.reserve(tuple->fields.size());
+    for (size_t i = 0; i < tuple->fields.size(); ++i) {
+      Expr field;
+      if (const auto* expr_tuple = expr.as<TupleNode>()) {
+        field = expr_tuple->fields[i];
+      } else {
+        field = TupleGetItem(expr, i);
+        field->struct_info_ = tuple->fields[i];
+      }
+      fields.push_back(TransformTupleLeaf(field, msg.NestedArray()[i], fvisitleaf));
```

Comment:
run msg.NestedArray() once outside the loop

Reply:
will fix this in the separate nestedmsg pr